features/steps/oauth.py

```diff
@@ -92,6 +105,90 @@ def step_impl(context):
     context.authorization_received = response
 
 
+@when('I ask for authorization')
+def step_impl(context):
+    context.code = context.oauth.request_authorization()
+
+
+@when('I exchange my grant code')
+def step_impl(context):
+    fields = {
+        'grant_type': 'authorization_code',
+        'code': context.code,
+        'client_id': context.vendor_config['auth']['client_id'],
+        'redirect_uri': context.vendor_config['auth']['redirect_uri'],
+    }
+
+    uris = fhir.get_oauth_uris(context.conformance)
+
+    auth = None
+    if context.vendor_config['auth'].get('confidential_client'):
+        auth = requests.auth.HTTPBasicAuth(
+            context.vendor_config['auth']['client_id'],
+            context.vendor_config['auth']['client_secret'],
+        )
+
+    response = requests.post(uris['token'],
+                             auth=auth,
+                             data=fields)
+
+    context.response = response
+
+
+@when('I exchange my grant code without the {field_name} field')
+def step_impl(context, field_name):
+    fields = {
+        'grant_type': 'authorization_code',
+        'code': context.code,
+        'client_id': context.vendor_config['auth']['client_id'],
+        'redirect_uri': context.vendor_config['auth']['redirect_uri'],
+    }
+
+    del fields[field_name]
+
+    uris = fhir.get_oauth_uris(context.conformance)
+
+    auth = None
+    if context.vendor_config['auth'].get('confidential_client'):
+        auth = requests.auth.HTTPBasicAuth(
+            context.vendor_config['auth']['client_id'],
+            context.vendor_config['auth']['client_secret'],
+        )
+
+    response = requests.post(uris['token'],
+                             auth=auth,
+                             data=fields)
+
+    context.response = response
+
+
+@when('I exchange my grant code with the following override')
+def step_impl(context):
+    fields = {
+        'grant_type': 'authorization_code',
+        'code': context.code,
+        'client_id': context.vendor_config['auth']['client_id'],
+        'redirect_uri': context.vendor_config['auth']['redirect_uri'],
+    }
+
+    fields.update(dict(context.table))
+
+    uris = fhir.get_oauth_uris(context.conformance)
+
+    auth = None
```

Comment:
Might abstract this pattern (find a token URL, determine whether confidential, add an auth header, make a request) or some part of it into function, since it repeats.